lib/ramble/ramble/reports.py

```diff
@@ -124,81 +141,239 @@ def filter_exp_results(experiments: list):
-    """Creates a Pandas DataFrame from the results dictionary to use for reports.
+def generate_result_index(experiments: list, all_vars=False, where_query=None):
+    """Creates an index from the results in the list of experiments
 
-    Transforms nested results dictionary into a flat dataframe. Each row equals
-    one FOM from one context of one experiment, with columns including
-    associated experiment variables (except paths and commands).
+    Index format is:
+    {
+        "applications": {
+            application_name: {
+                workload: {
+                    "Contexts": set(),
+                    "FOMs": set(),
+                    "Template Variables": set(),
+                }
+            }
+        }
+        "modifiers": {
+            modifier_name: {
+                "Contexts": set(),
+                "FOMs": set(),
+            }
+        (all other object types)
+    }
     """
-    filtered_exps = filter_exp_results(results["experiments"])
+    result_index: Dict[str, dict] = {}
+    for obj_name in OBJECT_NAMES.values():
+        result_index[obj_name] = {}
+
+    # Exclude dict items that aren't needed for analysis from all_vars, mainly paths and commands
+    dir_regex = r"_dir$"
+    path_regex = r"_path$"
+    vars_to_ignore = {
+        keywords.batch_submit,
+        keywords.log_file,
+        "command",
+        "execute_experiment",
+        "experiment_hash",
+        "experiment_status",
+        "name",
+        "RAMBLE_STATUS",
+        "CONTEXTS",
+        "RAMBLE_VARIABLES",
+        "RAMBLE_RAW_VARIABLES",
+        "SOFTWARE",
+        "TAGS",
+        "VARIANTS",
+        "EXPERIMENT_CHAIN",
+        "SUCCESS_CRITERIA",
+    }
+
+    template_patterns: Dict[str, dict] = {}
+    # First unnest dictionaries
+    for exp in experiments:
+        if exp["application_name"] not in result_index["applications"]:
+            result_index["applications"][exp["application_name"]] = {}
+        app_dict = result_index["applications"][exp["application_name"]]
+
+        if exp["workload_name"] not in app_dict:
+            app_dict[exp["workload_name"]] = {
+                "Contexts": set(),
+                "FOMs": set(),
+                "Template Variables": set(),
+            }
+        if exp["application_name"] not in template_patterns:
+            template_patterns[exp["application_name"]] = {}
+        if exp["workload_name"] not in template_patterns[exp["application_name"]]:
+            template_patterns[exp["application_name"]][exp["workload_name"]] = set()
+
+        if all_vars:
+            if "All Variables" not in app_dict[exp["workload_name"]]:
+                app_dict[exp["workload_name"]]["All Variables"] = set()
+            for var_name in exp.keys():
+                # Exclude non-variables
+                if var_name in vars_to_ignore:
+                    continue
+                if re.search(dir_regex, var_name):
+                    continue
```

Comment:
In practice is this always actually ends with? I wonder if we should prefer regular python over regex